File: python/refrase/models/nemotron.py
"""Nemotron model family adapter (9B, 12B VL, 30B)."""

from refrase.helpers import add_json_reinforcement, simplify_prompt
from refrase.models._base import BaseAdapter
from refrase.types import Change, ModelFamily, ModelInfo, TaskType


class NemotronAdapter(BaseAdapter):
    """Adapter for NVIDIA Nemotron models.

    Key adaptations:
    - 9B/12B: Tier 3 JSON prompt fallback, /think toggle, simplified prompts
    - 30B: Standard adaptation with tool use support
    """

    def _get_think_prefix(self, model_variant: str) -> str:
        """Determine thinking mode prefix for small models."""
        if "nothink" in model_variant:
            return "/no_think\n"
        return "/think\n"

    def adapt_system(
        self,
        system: str,
        task: TaskType,
        model_variant: str = "",
    ) -> str:
        is_9b = "9b" in model_variant
        is_12b = "12b" in model_variant
        is_small = is_9b or is_12b

        if is_small:
            think_prefix = self._get_think_prefix(model_variant)
            adapted = think_prefix + simplify_prompt(system, max_steps=3)
            adapted = add_json_reinforcement(adapted, tier3=True)
        else:
            adapted = system
            adapted = add_json_reinforcement(adapted)

        return adapted

    def adapt_user(
        self,
        user: str,
        task: TaskType,
        model_variant: str = "",
    ) -> str:
        is_small = "9b" in model_variant or "12b" in model_variant
        if is_small:
            prefix = "/no_think\n" if "nothink" in model_variant else "/think\n"
            return prefix + simplify_prompt(user, max_steps=3)
        return user

    def get_changes(self, task: TaskType, model_variant: str = "") -> list[Change]:
        is_small = "9b" in model_variant or "12b" in model_variant

        if is_small:
            return [
                Change(
                    rule="thinking-mode",
                    description=f"Added {self._get_think_prefix(model_variant).strip()} prefix",
                    evidence="Nemotron 9B/12B support /think and /no_think toggles",
                    impact="Optimized reasoning mode",
                ),
                Change(
                    rule="simplification",
                    description="Simplified prompt for small model capacity",
                    evidence="9B/12B have limited attention layers",
                    impact="Reduced prompt length",
                ),
                Change(
                    rule="tier3-reinforcement",
                    description="Added extra-strong JSON-only instructions",
                    evidence="9B/12B use JSON-prompt fallback (Tier 3)",
                    impact="Improved schema compliance",
                ),
            ]
        return [
            Change(
                rule="json-reinforcement",
                description="Added JSON output compliance instructions",
                evidence="Explicit JSON instructions improve schema adherence",
                impact="More reliable structured output",
            ),
        ]

    def get_model_info(self) -> ModelInfo:
        return ModelInfo(
            family=ModelFamily.NEMOTRON,
            description="NVIDIA Nemotron models (9B, 12B VL, 30B)",
            adaptations=[
                "9B/12B: thinking mode control",
                "9B/12B: Tier 3 JSON reinforcement",
                "9B/12B: prompt simplification",
                "30B: standard JSON reinforcement",
            ],
        )

File: typescript/src/models/nemotron.ts
import type { BaseAdapter } from "./base.js";
import type { Change, ModelInfo, TaskType } from "../types.js";
import { addJsonReinforcement, simplifyPrompt } from "../helpers/index.js";

export class NemotronAdapter implements BaseAdapter {
  private getThinkPrefix(modelVariant: string): string {
    if (modelVariant.includes("nothink")) return "/no_think\n";
    return "/think\n";
  }

  adaptSystem(system: string, _task: TaskType, modelVariant = ""): string {
    const isSmall =
      modelVariant.includes("9b") || modelVariant.includes("12b");

    if (isSmall) {
      const thinkPrefix = this.getThinkPrefix(modelVariant);
      let adapted = thinkPrefix + simplifyPrompt(system, 3);
      return addJsonReinforcement(adapted, true);
    }

    return addJsonReinforcement(system);
  }

  adaptUser(user: string, _task: TaskType, modelVariant = ""): string {
    const isSmall =
      modelVariant.includes("9b") || modelVariant.includes("12b");
    if (isSmall) {
      const prefix = modelVariant.includes("nothink")
        ? "/no_think\n"
        : "/think\n";
      return prefix + simplifyPrompt(user, 3);
    }
    return user;
  }

  getChanges(_task: TaskType, modelVariant = ""): Change[] {
    const isSmall =
      modelVariant.includes("9b") || modelVariant.includes("12b");

    if (isSmall) {
      return [
        {
          rule: "thinking-mode",
          description: `Added ${this.getThinkPrefix(modelVariant).trim()} prefix`,
          evidence: "Nemotron 9B/12B support /think and /no_think toggles",
          impact: "Optimized reasoning mode",
        },
        {
          rule: "simplification",
          description: "Simplified prompt for small model capacity",
          evidence: "9B/12B have limited attention layers",
          impact: "Reduced prompt length",
        },
        {
          rule: "tier3-reinforcement",
          description: "Added extra-strong JSON-only instructions",
          evidence: "9B/12B use JSON-prompt fallback (Tier 3)",
          impact: "Improved schema compliance",
        },
      ];
    }
    return [
      {
        rule: "json-reinforcement",
        description: "Added JSON output compliance instructions",
        evidence: "Explicit JSON instructions improve schema adherence",
        impact: "More reliable structured output",
      },
    ];
  }

  getModelInfo(): ModelInfo {
    return {
      family: "nemotron",
      description: "NVIDIA Nemotron models (9B, 12B VL, 30B)",
      adaptations: [
        "9B/12B: thinking mode control",
        "9B/12B: Tier 3 JSON reinforcement",
        "9B/12B: prompt simplification",
        "30B: standard JSON reinforcement",
      ],
    };
  }
}
